tree/tree/src/TChain.cxx

```diff
@@ -575,6 +577,31 @@ Int_t TChain::AddFile(const char* name, Long64_t nentries /* = TTree::kMaxEntrie
    return 1;
 }
 
+
+////////////////////////////////////////////////////////////////////////////////
+/// Add a new list of files to this chain.
+///
+/// Filename formats are similar to TChain::Add. Wildcards are not
+/// applied. urls may also contain query and fragment identifiers
+/// where the tree name can be specified in the url fragment.
+
+Int_t TChain::AddFilelist(const char* filelist, Long64_t nentries /* = TTree::kMaxEntries */, const char* tname /* = "" */)
+{
+   if (filelist == 0 || filelist[0] == '\0') {
+      Error("AddFilelist", "No file name; no files connected");
+      return 0;
+   }
+
+   Int_t sum_results = 0;
+   std::vector<std::string> input_file_list = ROOT::ReadFilelist(filelist);
+   for (auto f: input_file_list) {
+     Int_t result = AddFile(f.c_str(), nentries, tname);
```

Comment:
On this line the function assumes that `nentries` and `tname`  are the same value for all files in `filelist`. I would say this is not true in general. Rather, the function signature should be
```cpp
Int_t TChain::AddFilelist(const char* filelist, const std::vector<Long64_t> &nentries, const std::vector<const char*> &tnames)
```
But this would also mean that the caller is already aware of the total number of files in the `filelist` parameter. This would have the advantage that at least we could `reserve` the size of the vector `input_file_list` in the `ReadFileList` function. But at the same time, it brings a non trivial complication for the user

Reply:
Well, when processing (applying a method) a bulk of elements, the underlying idea IMHO is that on each element the same thing is applied .. so, not only that is not within the scope of this PR, but even generally i'm reluctant to add complexity for individual customization per element. With this PR i just intended a simple `for_each file in filelist: add file to chain`. If other complex behaviors are desired, i would say to be added after this, and when the requirements are clear (personally for my usage i see no use for this) those who need it know best what their requirements are and can add their implementation afterwards

Reply:
I would say having a different number of entries per tree is far more common that having a set of files with the exact same number of entries? I don't think it is a complication but just the most common situation. If we want to start easily, than I would change the signature to 
```cpp
Int_t TChain::AddFilelist(const char* filelist)
```
and in the for loop just write
```cpp
   for (const auto &f: input_file_list) {
     Int_t result = AddFile(f.c_str());
```

Reply:
well, but while personally i encode the tree name in the actual file, there could be cases where tree name is specified in the 3rd argument... then maybe, following what you suggest, should i change signature to (filelist, tname) and within the loop to do AddFile(f, MaxEntries, tname)? Would it be ok this way?

Reply:
If all trees in the chain have the same name, then you don't need to do it. For example,
```cpp
TChain c{"Events"};
std::vector<std::string> filenames = {"file1.root","file2.root","file3.root"};
for (const auto &name: filenames){
    c.AddFile(name.c_str());
}
```
This will add the three files in the list to the chain and it will be assumed that the trees inside the files share a common name `Events` given in the TChain constructor.

So if you have a list of file names containing trees with the same name, you don't need to add the extra `tname` parameter to the function signature.

On the other hand, if each file in your list might contain a tree with a different name (e.g. `file1.root, Events1`, `file2.root, Events2`, `file3.root, Events3`) than we go back to my original comment and the function signature should have a parameter like `const std::vector<std::string> &treenames` and you would need to traverse the two vectors in lock-step in the loop

Reply:
ok, then i will resolve the conversation. thanks a lot!